src/Microsoft.ML.Core/Environment/HostEnvironmentBase.cs

```diff
@@ -96,36 +96,47 @@ internal interface IMessageSource
     internal abstract class HostEnvironmentBase<TEnv> : ChannelProviderBase, IHostEnvironment, IChannelProvider
         where TEnv : HostEnvironmentBase<TEnv>
     {
+        [BestFriend]
+        internal void CancelExecutionHosts()
+        {
+            lock (_cancelLock)
+            {
+                foreach (var child in _children)
+                    if (child.TryGetTarget(out IHost host))
+                        if (host is ICancelableHost cancelableHost)
+                            cancelableHost.CancelExecution();
+
+                _children.Clear();
+            }
+        }
+
         /// <summary>
         /// Base class for hosts. Classes derived from  <see cref="HostEnvironmentBase{THostEnvironmentBase}"/> may choose
         /// to provide their own host class that derives from this class.
         /// This encapsulates the random number generator and name information.
         /// </summary>
-        public abstract class HostBase : HostEnvironmentBase<TEnv>, IHost
+        public abstract class HostBase : HostEnvironmentBase<TEnv>, ICancelableHost
         {
             public override int Depth { get; }
 
             public Random Rand => _rand;
 
-            // We don't have dispose mechanism for hosts, so to let GC collect children hosts we make them WeakReference.
-            private readonly List<WeakReference<IHost>> _children;
-
             public HostBase(HostEnvironmentBase<TEnv> source, string shortName, string parentFullName, Random rand, bool verbose)
                 : base(source, rand, verbose, shortName, parentFullName)
             {
                 Depth = source.Depth + 1;
-                _children = new List<WeakReference<IHost>>();
             }
 
-            public void StopExecution()
+            public void CancelExecution()
```

Comment:
This appears to be duplicated with the base `CancelExecutionHosts`. Can we just use the base implementation?